### ISSUE TITLE
Lua improvements: Allow multi-match, stream direction, catch-all also on non http streams

### ISSUE THREAD

**Issue by pointhi** [arkime/arkime#1549]
Based on the doc from https://github.com/arkime/arkime/tree/master/capture/plugins/lua I see multiple issues to implement an efficient tagging feature which is aware of the stream direction. E.g. find all matches of an regex in the TCP stream and put them into a custom field for efficient search.

* [ ] `MolochSession.register_body_feed` only works for http and smtp. What about other tcp/udp streams? Registration of classifiers is based on a pattern, but I want to match on the streams, and not only the first packet.
* [ ] It seems impossible to extract the direction of a match in MolochData using register_body_feed?
* [ ] Find all matches of a regex in a stream is not supported. (from what I see in the doc, only one match is returned)

It seems I could hack this feature into a script using e.g. `session:register_parser` and custom regex matching, but this does not look like an intended solution. My current code snippet which works at least for the basic stuff:

```lua
function regex_detection(session, data)
    local matched, match = data:pcre_match(regex)

    -- TODO: multiple matches, direction of match, non HTTP streams?
    if matched then
        session:add_string("custom.field", match)
    end
end

regex = MolochData.pcre_create([[\bREGEX[0-9]{5}\b]])
MolochSession.register_body_feed("http", "regex_detection")
```

**Comment by awick** [comment]
> * [ ] It seems impossible to extract the direction of a match in MolochData using register_body_feed?

ya, looks like that could be added like the other callbacks have

> * [ ]  `MolochSession.register_body_feed` only works for http and smtp. What about other tcp/udp streams? Registration of classifiers is based on a pattern, but I want to match on all the streams, and not only the first packet.

Currently that is a limit of how Arkime works, even for C parsers. Can you give a real world example

> * [ ]  Find all matches of a regex in a stream is not supported. (from what I see in the doc, only one match is returned)

Arkime doesn't really have the whole stream, it has callbacks as it gets the data. Or do you mean you want to find all matches in a single chunk of data?

**Comment by pointhi** [comment]
Yes, I want to find all matches in a stream. Similar to yara rules, but to extract the matches as well and makes them available in the database, as well as differ them by the direction on the stream.

**Comment by awick** [comment]
So again, a real world example will help, since I'm not really sure what you are trying to do or if it fits with Arkime goals.

**Comment by pointhi** [comment]
We use Akime for CTF games, where a lot of data is flowing around (including so called flags which are defined using a regular expression). We want to find them, differ between ingoing/outgoing flags as well as send them to a secondary system where monitoring, etc. is happening.

Right now, we use yara rules to find flags, and an external python service which downloads each tx/rx stream separatly where the yara tag is set, searches for those flags again and sends them to the secondary service. This is a pretty brittle solution because sometimes only metadata is available, but not the streams because they are still not written to disk (or due to configuration issues). In short, it is easy to hang this system up because Akime did not expose any way to get tagged streams reliable into a secondary system for analysis. The Lua interface on the other hand would provide a pretty stable solution.
